File: packages/python-sdk/src/urisocial/resources/publishing.py
"""Publishing resource"""

from typing import TYPE_CHECKING, Any, Dict, List, Optional

if TYPE_CHECKING:
    from ..http_client import HTTPClient


class PublishingResource:
    """Resource for the content approval / scheduling / publishing workflow"""

    def __init__(self, http: "HTTPClient"):
        self._http = http

    def approve(
        self,
        draft_ids: List[str],
        schedule_option: str = "save_draft",
        scheduled_datetime: Optional[str] = None,
        approval_notes: Optional[str] = None,
    ) -> dict:
        """
        Approve content drafts, optionally publishing or scheduling them.

        Args:
            draft_ids: IDs of drafts to approve
            schedule_option: One of 'immediate', 'schedule', 'save_draft'
            scheduled_datetime: ISO 8601 datetime string, required when schedule_option='schedule'
            approval_notes: Optional notes to attach to the approval

        Example:
            >>> result = client.publishing.approve(
            ...     draft_ids=['draft-123'],
            ...     schedule_option='immediate',
            ... )
        """
        body: Dict[str, Any] = {
            "draft_ids": draft_ids,
            "schedule_option": schedule_option,
        }
        if scheduled_datetime is not None:
            body["scheduled_datetime"] = scheduled_datetime
        if approval_notes is not None:
            body["approval_notes"] = approval_notes
        return self._http.post("/social-media/approve", json=body)

    def deny(
        self,
        draft_ids: List[str],
        denial_reason: str,
        request_regeneration: bool = False,
    ) -> dict:
        """
        Deny/reject drafts in the approval workflow.

        Args:
            draft_ids: IDs of drafts to deny
            denial_reason: Reason for denial
            request_regeneration: Whether to trigger regeneration of the denied drafts
        """
        return self._http.post(
            "/social-media/deny",
            json={
                "draft_ids": draft_ids,
                "denial_reason": denial_reason,
                "request_regeneration": request_regeneration,
            },
        )

    def refine(self, draft_id: str, refinements: Dict[str, Any]) -> dict:
        """
        Request refinements to a draft.

        Args:
            draft_id: ID of the draft to refine
            refinements: Specific changes to make
        """
        return self._http.put(
            "/social-media/refine",
            json={"draft_id": draft_id, "refinements": refinements},
        )

    def schedule(
        self,
        draft_ids: List[str],
        scheduled_datetime: str,
        timezone: str = "UTC",
    ) -> dict:
        """
        Schedule drafts for future publishing.

        Args:
            draft_ids: IDs of drafts to schedule
            scheduled_datetime: ISO 8601 datetime string for publishing
            timezone: IANA timezone name (default 'UTC')

        Example:
            >>> result = client.publishing.schedule(
            ...     draft_ids=['draft-123'],
            ...     scheduled_datetime='2026-12-25T10:00:00Z',
            ... )
        """
        return self._http.post(
            "/social-media/schedule",
            json={
                "draft_ids": draft_ids,
                "scheduled_datetime": scheduled_datetime,
                "timezone": timezone,
            },
        )

    def list_scheduled(self) -> dict:
        """
        Get scheduled posts.

        Example:
            >>> scheduled = client.publishing.list_scheduled()
        """
        return self._http.get("/social-media/scheduled")

    def publish_scheduled(self, draft_ids: List[str]) -> dict:
        """
        Manually trigger publishing of already-scheduled drafts.

        Args:
            draft_ids: IDs of scheduled drafts to publish immediately
        """
        return self._http.post(
            "/social-media/publish-scheduled", json={"draft_ids": draft_ids}
        )

    def cancel_scheduled(self, draft_id: str) -> dict:
        """
        Cancel a scheduled post.

        Args:
            draft_id: ID of the scheduled draft to cancel

        Example:
            >>> client.publishing.cancel_scheduled('draft-123')
        """
        return self._http.post(f"/social-media/drafts/{draft_id}/unschedule")

File: packages/typescript-sdk/src/resources/publishing.ts
import { HTTPClient } from '../client/http';
import { PublishRequest, PublishResult, Platform } from '../types';

export interface ScheduleRequest {
  draft_ids: string[];
  scheduled_datetime: string; // ISO 8601 format
  timezone?: string;
}

export interface ScheduledPost {
  id: string;
  draft_id: string;
  platforms: Platform[];
  scheduled_for: string;
  status: 'pending' | 'published' | 'failed';
  created_at: string;
}

export interface ApprovalRequest {
  draft_ids: string[];
  schedule_option: 'save_draft' | 'immediate' | 'schedule';
  scheduled_datetime?: string;
  approval_notes?: string;
}

export interface DenialRequest {
  draft_ids: string[];
  denial_reason: string;
  request_regeneration?: boolean;
}

export interface RefinementRequest {
  draft_id: string;
  refinements: Record<string, any>; // Specific changes to make
}

export class PublishingResource {
  constructor(private http: HTTPClient) {}

  /**
   * Publish approved drafts immediately
   * Approves and publishes selected drafts to their respective platforms
   * @param request - Approval request with draft IDs and scheduling options
   */
  async approve(request: ApprovalRequest): Promise<{
    success: boolean;
    published_drafts: any[];
    scheduled_drafts?: any[];
  }> {
    return this.http.post('/social-media/approve', request);
  }

  /**
   * Publish a draft to social media platforms
   * @deprecated Use approve() for better approval workflow support
   */
  async publish(request: PublishRequest): Promise<{
    results: PublishResult[];
    overall_status: 'success' | 'partial' | 'failed';
  }> {
    return this.http.post('/api/v1/publish', request);
  }

  /**
   * Schedule drafts for future publishing
   * @param request - Schedule configuration
   */
  async schedule(request: ScheduleRequest): Promise<{
    success: boolean;
    scheduled_posts: ScheduledPost[];
  }> {
    return this.http.post('/social-media/schedule', request);
  }

  /**
   * Get all scheduled posts
   */
  async listScheduled(): Promise<{
    scheduled_posts: ScheduledPost[];
  }> {
    return this.http.get('/social-media/scheduled');
  }

  /**
   * Publish scheduled posts manually (trigger scheduled posts immediately)
   * @param draftIds - Array of draft IDs to publish
   */
  async publishScheduled(draftIds: string[]): Promise<{
    success: boolean;
    results: PublishResult[];
  }> {
    return this.http.post('/social-media/publish-scheduled', { draft_ids: draftIds });
  }

  /**
   * Deny/reject drafts in approval workflow
   * @param request - Denial request with reason
   */
  async deny(request: DenialRequest): Promise<{
    success: boolean;
    denied_drafts: string[];
  }> {
    return this.http.post('/social-media/deny', request);
  }

  /**
   * Request refinements to a draft
   * @param request - Refinement request with specific changes
   */
  async refine(request: RefinementRequest): Promise<{
    success: boolean;
    refined_draft: any;
  }> {
    return this.http.put('/social-media/refine', request);
  }

  /**
   * Cancel a scheduled post
   * @deprecated Use drafts.unschedule() instead
   */
  async cancelScheduled(scheduledId: string): Promise<{ success: boolean }> {
    return this.http.delete(`/api/v1/publish/scheduled/${scheduledId}`);
  }
}
